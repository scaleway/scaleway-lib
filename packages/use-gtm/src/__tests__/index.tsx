import { fireEvent, renderHook } from '@testing-library/react'
import mockdate from 'mockdate'
import type { ReactNode } from 'react'
import type { SendGTM } from '..'
import GTMProvider, { useGTM } from '..'
import type { GTMProviderProps } from '../useGTM'

const defaultEvents = {
  sampleEvent: (sendGTM?: SendGTM) => (extraValue: string) => {
    sendGTM?.({
      event: 'sampleEvent',
      extra: extraValue,
    })
  },
}

type DefaultEvents = typeof defaultEvents

const wrapper =
  ({
    id,
    events,
    environment,
    onLoadError,
  }: Omit<GTMProviderProps<DefaultEvents>, 'children'>) =>
  ({ children }: { children: ReactNode }) => (
    <GTMProvider
      id={id}
      events={events}
      environment={environment}
      onLoadError={onLoadError}
    >
      {children}
    </GTMProvider>
  )

describe('GTM hook', () => {
  beforeEach(() => {
    mockdate.set('4/13/2021')
  })

  afterEach(() => {
    mockdate.reset()
    document.head.innerHTML = ''
    window.dataLayer = undefined
    jest.restoreAllMocks()
  })

  it('useGTM should not be defined without GTMProvider', () => {
    const orignalConsoleError = console.error
    console.error = jest.fn

    try {
      renderHook(() => useGTM())
    } catch (error) {
      expect((error as Error).message).toBe(
        'useGTM must be used within a GTMProvider',
      )
    }

    console.error = orignalConsoleError
  })

  it('Provider should call onLoadError if script fail to load', () => {
    renderHook(() => useGTM<DefaultEvents>(), {
      wrapper: wrapper({
        id: 'testId',
      }),
    })

    expect(document.head.innerHTML).toMatchSnapshot()
  })

  it('Provider should not load when no id is provided', () => {
    renderHook(() => useGTM<DefaultEvents>(), {
      wrapper: wrapper({}),
    })

    expect(document.head.innerHTML).toMatchSnapshot()
  })

  it('Provider should load when id is provided', () => {
    renderHook(() => useGTM<DefaultEvents>(), {
      wrapper: wrapper({
        id: 'testId',
      }),
    })

    expect(document.head.innerHTML).toMatchSnapshot()
  })

  it('Provider should load when id and environment is provided', () => {
    renderHook(() => useGTM<DefaultEvents>(), {
      wrapper: wrapper({
        environment: {
          auth: 'gtm',
          preview: 'world',
        },
        id: 'testId',
      }),
    })

    expect(document.head.innerHTML).toMatchSnapshot()
  })

  it('Provider should load env when environment auth is missing', () => {
    renderHook(() => useGTM<DefaultEvents>(), {
      wrapper: wrapper({
        // @ts-expect-error we test a failing case
        environment: {
          preview: 'world',
        },
        id: 'testId',
      }),
    })

    expect(document.head.innerHTML).toMatchSnapshot()
  })

  it('Provider should load with events when provided', () => {
    const { result } = renderHook(() => useGTM<DefaultEvents>(), {
      wrapper: wrapper({
        events: defaultEvents,
        id: 'testId',
      }),
    })

    expect(result.current.events.sampleEvent('test')).toBe(undefined)
    expect(window.dataLayer).toMatchSnapshot()
    // @ts-expect-error if type infering works this should be an error
    expect(result.current.events.sampleEvent()).toBe(undefined)
  })

  it('Provider should load onLoadError when script fail to load', () => {
    const onLoadError = jest.fn()

    renderHook(() => useGTM<DefaultEvents>(), {
      wrapper: wrapper({
        id: 'testId',
        onLoadError,
      }),
    })

    const script = document.querySelector(
      `script[src="https://www.googletagmanager.com/gtm.js?id=testId"]`,
    ) as Element
    fireEvent.error(script)
    expect(onLoadError).toHaveBeenCalledTimes(1)
  })
})
