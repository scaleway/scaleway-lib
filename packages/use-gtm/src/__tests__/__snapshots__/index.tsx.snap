// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GTM hook Provider should call onLoadError if script fail to load 1`] = `
"<script src="https://www.googletagmanager.com/gtm.js?id=testId"></script><script>window.dataLayer = window.dataLayer || [];</script><script>(function(w,d,s,l,i){w[l]=w[l]||[];
    w[l].push({'gtm.start': new Date().getTime(),event:'gtm.js'});
    var f=d.getElementsByTagName(s)[0],j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';
    j.async=true;j.src='https://www.googletagmanager.com/gtm.js?id='+i+dl+'';

    j.addEventListener('error', function() {
        var _ge = new CustomEvent('gtm_loading_error', { bubbles: true });
        d.dispatchEvent(_ge);
    });

    f.parentNode.insertBefore(j,f);
  })(window,document,'script','dataLayer','testId');</script>"
`;

exports[`GTM hook Provider should load env when environment auth is missing 1`] = `
"<script src="https://www.googletagmanager.com/gtm.js?id=testId"></script><script>window.dataLayer = window.dataLayer || [];</script><script>(function(w,d,s,l,i){w[l]=w[l]||[];
    w[l].push({'gtm.start': new Date().getTime(),event:'gtm.js'});
    var f=d.getElementsByTagName(s)[0],j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';
    j.async=true;j.src='https://www.googletagmanager.com/gtm.js?id='+i+dl+'';

    j.addEventListener('error', function() {
        var _ge = new CustomEvent('gtm_loading_error', { bubbles: true });
        d.dispatchEvent(_ge);
    });

    f.parentNode.insertBefore(j,f);
  })(window,document,'script','dataLayer','testId');</script>"
`;

exports[`GTM hook Provider should load when id and environment is provided 1`] = `
"<script src="https://www.googletagmanager.com/gtm.js?id=testId&amp;gtm_auth=gtm&amp;gtm_preview=world&amp;gtm_cookies_win=x"></script><script>window.dataLayer = window.dataLayer || [];</script><script>(function(w,d,s,l,i){w[l]=w[l]||[];
    w[l].push({'gtm.start': new Date().getTime(),event:'gtm.js'});
    var f=d.getElementsByTagName(s)[0],j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';
    j.async=true;j.src='https://www.googletagmanager.com/gtm.js?id='+i+dl+'&gtm_auth=gtm&gtm_preview=world&gtm_cookies_win=x';

    j.addEventListener('error', function() {
        var _ge = new CustomEvent('gtm_loading_error', { bubbles: true });
        d.dispatchEvent(_ge);
    });

    f.parentNode.insertBefore(j,f);
  })(window,document,'script','dataLayer','testId');</script>"
`;

exports[`GTM hook Provider should load when id is provided 1`] = `
"<script src="https://www.googletagmanager.com/gtm.js?id=testId"></script><script>window.dataLayer = window.dataLayer || [];</script><script>(function(w,d,s,l,i){w[l]=w[l]||[];
    w[l].push({'gtm.start': new Date().getTime(),event:'gtm.js'});
    var f=d.getElementsByTagName(s)[0],j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';
    j.async=true;j.src='https://www.googletagmanager.com/gtm.js?id='+i+dl+'';

    j.addEventListener('error', function() {
        var _ge = new CustomEvent('gtm_loading_error', { bubbles: true });
        d.dispatchEvent(_ge);
    });

    f.parentNode.insertBefore(j,f);
  })(window,document,'script','dataLayer','testId');</script>"
`;

exports[`GTM hook Provider should load with events when provided 1`] = `
[
  {
    "event": "gtm.js",
    "gtm.start": 1618272000000,
  },
  {
    "event": "sampleEvent",
    "extra": "test",
  },
]
`;

exports[`GTM hook Provider should not load when no id is provided 1`] = `""`;
